### PR TITLE
introduce `on_points`, `on_tuples`, ...

### PR DESCRIPTION
**Pull request by ThomasBreuer**
This is a first attempt to introduce action functions.
Once it is clear that this idea is o.k.,
more such functions have to follow, more methods have to be provided
(for example `on_right` for right cosets and group elements),
and then functions for orbits, stabilizers, induced permutations, etc.
can be added.

### CODE REVIEW

**Comment by thofma** [comment]
Regarding the syntax, we could additionally overload directly `x^g` or `g(x)` respectively `x * g` and `g * x`. But this would just be syntactic sugar.

**Comment by fingolfin** [comment]
Thank you! (if you haven't already, please also see PR #119, which I just merged (and which is super boring and nothing much happens, I thought it'd be good for you to be aware), and issue #108, which still lacks details ).

I envision that we will do two things in parallel:
1. provide `on_FOOBAR` functions like this PR does
2. provide a more Julia-like API which exploits the richer types for datastructures:  when considering `[[1,2], [3,4]]^g` in GAP, this is ambiguous, hence we have `OnSets`, `OnTuples`, `OnSetsSets`, `OnSetsTuples` etc.. In Julia, though, we can use the type: `Tuple` vs. `Set` vs. `BitSet` etc. (and we probably will want some extra types, too: `Set` is great but for small sets with just a single digits of entries, it might be more efficient to use a "sorted vector" that works more like what GAP does; see e.g. https://github.com/tpapp/SortedVectors.jl or https://github.com/JuliaIntervals/IntervalOptimisation.jl/blob/master/src/SortedVectors.jl ; but actually we could just have our own super simple one which basically just adds a "tag" to the vector that it is meant to be sorted and should be transformed as such by permutations).

In the second approach, what Tommy wrote absolutely makes sense, as we can "automatically" decide for the "correct" action. In fact, I'd just implement the actions as such.

(However, I am not quite sure how to implement `g(x)` in terms of `x^g`, other than perhaps by inserting an inversion somewhere; that's about the only generic way to turn a right action into a left action, no?

Anyway, I'll look at this PR after our status meeting :-)

File: src/Groups/action.jl [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+    gapv = GAP.julia_to_gap(v)
+    img = GAP.Globals.OnRight(gapv, x.X)
+    return Vector{T}(img)
+end
+
+
+"""
+    on_tuples(tuple::GAP.GapObj, x::GAPGroupElem)
+    on_tuples(tuple::Vector{T}, x::GAPGroupElem) where T
+    on_tuples(tuple::T, x::GAPGroupElem) where T <: Tuple
```
Comment (fingolfin) [inline src/Groups/action.jl:77]: How about this instead:
```suggestion
    on_tuples(tuple::Tuple{Vararg{T}}, x::GAPGroupElem) where T
```

Going beyond this (and perhaps beyond this PR): as mentioned already, for tuples, it'd also be natural to just define a `^` method directly, as it is "clear" how to act on that. Just as I'd argue that the default action for a vector ought to be OnRight, for a `Set` or `BitSet` OnSets etc. -- and then nested, i.e. a `Set{Set{Int}}` suggest `OnSetsSets`, while `Set{Tuple{Vararg{Int8}}}` suggests `OnSetsTuples` etc.

File: src/Groups/action.jl [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+GAP: (2,3)
+
+julia> on_points(1, g[1])
+2
+
+```
+"""
+on_points(pnt::GAP.GapObj, x::GAPGroupElem) = GAP.Globals.OnPoints(pnt, x.X)
+
+on_points(pnt::GAPGroupElem, x::GAPGroupElem) = inv(x) * pnt * x
```
Comment (fingolfin) [inline src/Groups/action.jl]: That's inefficient, e.g. for permutations, conjugation can be computed much more efficiently than that. So I think it really should be this:
```suggestion
on_points(pnt::GAPGroupElem, x::GAPGroupElem) = pnt ^ x
```

File: src/Groups/action.jl [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+##  on `GAP.GapObj` objects to the corresponding GAP action,
+##  and to implement the action on native Julia objects case by case.
+
+export on_points, on_right, on_tuples, on_sets, permuted
+
+
+"""
+    on_points(pnt::GAP.GapObj, x::GAPGroupElem)
+    on_points(pnt::GAPGroupElem, x::GAPGroupElem)
+    on_points(pnt::Int, x::PermGroupElem)
```
Comment (fingolfin) [inline src/Groups/action.jl]: Honest question: What is the purpose in having `on_points` over just using `:^`? I.e. where one passes `on_points` as argument, couldn't one just pass `^` instead? It's quite likely I am missing something, but what? Note that I occasionally wonder the same about GAP.

File: src/Groups/action.jl [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+
+on_points(pnt::GAPGroupElem, x::GAPGroupElem) = inv(x) * pnt * x
+
+on_points(pnt::Int, x::PermGroupElem) = GAP.Globals.OnPoints(pnt, x.X)
+
+
+"""
+    on_right(pnt::GAP.GapObj, x::GAPGroupElem)
+    on_right(pnt::GAPGroupElem, x::GAPGroupElem)
+    on_right(v::Vector{T}, x::MatrixGroupElem) where T
```
Comment (fingolfin) [inline src/Groups/action.jl]: As for `OnPoints` / `on_points`, what is the advantage of having `on_right`, as compared to just passing `:*` as "the action" whenever needed?

**Comment by fingolfin** [comment]
@ThomasBreuer I just realized that I actually already left some feedback here, so waiting for your reply here now :-)

**Comment by fingolfin** [comment]
@ThomasBreuer do you plan to update this PR?

**Comment by ThomasBreuer** [comment]
@fingolfin Yes.

File: src/Groups/action.jl (hunk 1 above)
Comment (ThomasBreuer) [reply src/Groups/action.jl:77]: In my variant, the tuple need not be homogeneous, and it is simpler to create the result object of the right type.

File: src/Groups/action.jl [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+   invx = inv(x)
+   return T(pnt[[i^invx for i in 1:length(pnt)]])
+end
+
+
+"""
+    on_indeterminates(f::GAP.GapObj, p::PermGroupElem)
+    on_indeterminates(f::Nemo.MPolyElem{T}, p::PermGroupElem) where T
+
+Returns the image of `f` under `p`, w.r.t. permuting the indeterminates
```
Comment (rfourquet) [inline src/Groups/action.jl:224]:
```suggestion
Return the image of `f` under `p`, w.r.t. permuting the indeterminates
```

Comment (ThomasBreuer) [reply src/Groups/action.jl:224]: Yes, thanks for the hint.

**Review by fingolfin — APPROVED** [review]
Looks OK to me. I still would like us to also install default methods for ^ for acing on `Tuple`, `Set`, etc. "in the usual way", but that doesn't have to hold up this PR. Also, we'll still need dedicated actions anyway. E.g. when acting on a tuple or vector, while I'd consider `on_tuples` the "default", of course one could also act by permuting the tuple entries (`on_positions`? `on_indices`?)

Also, this should be consolidated with the to-be-fleshed-out concept of G-sets, but again, no need to hold this up for now.

Merge this at will?

**Comment by fieker** [comment]
On Fri, Sep 04, 2020 at 01:41:43AM -0700, Max Horn wrote:
> @fingolfin approved this pull request.
> 
> Looks OK to me. I still would like us to also install default methods for ^ for acing on `Tuple`, `Set`, etc. "in the usual way", but that doesn't have to hold up this PR. Also, we'll still need dedicated actions anyway. E.g. when acting on a tuple or vector, while I'd consider `on_tuples` the "default", of course one could also act by permuting the tuple entries (`on_positions`? `on_indices`?)
> 
> Also, this should be consolidated with the to-be-fleshed-out concept of G-sets, but again, no need to hold this up for now.
> 
> Merge this at will?
> 
> 
Yep. at will....

**Comment by ThomasBreuer** [comment]
@fingolfin I do not understand your latest comment. The default methods for `^` had been added in the updated version of the pull request, and I had added also an explanation at the beginning of the file why one cannot use `^` in all situations --eventually this has to become part of the documentation, but currently all this is work in progress.

**Comment by fingolfin** [comment]
@ThomasBreuer Sorry for the confusion. Apparently I was looking at an outdated version of the PR. I've force-reloaded and now see those `^` methods. I'll go over this PR once again.

**Comment by fingolfin** [comment]
Closing and re-opening to re-run the tests

**Pull request merged**